Add "Create a new campaign" modal

ecgan opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/34909. Previous draft PR and discussions are in PR https://github.com/woocommerce/woocommerce/pull/36130.

This PR implements the "Create a new campaign" modal.

![image](https://user-images.githubusercontent.com/417342/209197516-ec6ccf0e-3829-4de1-b550-b82f3d171af6.png)

![image](https://user-images.githubusercontent.com/417342/209197676-863e91fd-e2f8-4a5e-919a-ef8b34473702.png)

The implemented modal here is different from the figma designs in https://github.com/woocommerce/woocommerce/issues/34909. Specifically, there are no borders within the modal component. This is to be more consistent with the [WordPress Modal component](https://wordpress.github.io/gutenberg/?path=/story/components-modal--default) which no longer has border under the modal title since PR https://github.com/WordPress/gutenberg/pull/40781.

The new campaign types data is loaded from `GET /marketing/campaign-types` API. The API is called when the Campaigns card is rendered, so that when users click on the "Create new campaign" button, the data is already there and no additional waiting is required.

### How to test the changes in this Pull Request:

<!-- Please include detailed instructions on how these changes can be tested, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [x] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

Pre-conditions: 

- Enable WooCommerce Multichannel Marketing experience in WooCommerce Settings > Advanced > Features page: `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`
- Installed and activated a marketing channel, such as: 
	- Google Listings and Ads (Note: you may need to use a filter, see PR https://github.com/woocommerce/google-listings-and-ads/pull/1904).
	- Pinterest for WooCommerce (see PR https://github.com/woocommerce/pinterest-for-woocommerce/pull/705).

1. Go to WooCommerce Marketing page: `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`
2. Click on the "Create new campaign" button in the Campaigns card.
3. You should see the modal like the screenshot above.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [x] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

ecgan committed:
Add CampaignsCardHeader component with modal.

ecgan committed:
Remove unneeded imports in RecommendedChannelsList.

ecgan committed:
Move RecommendedChannelsList to shared components directory.

ecgan committed:
Move useRecommendedChannels to shared hooks directory.

ecgan committed:
Display recommended channels in create campaign modal.

ecgan committed:
Hide the recommended channels section if it is empty.

ecgan committed:
Display new campaign types in create campaign modal.

ecgan committed:
Make create campaign button open in new tab depending on createUrl origin.

ecgan committed:
Code refactor with CreateNewCampaignModal.

ecgan committed:
Better code clarity with woocommerce-marketing-campaigns-table class name.

ecgan committed:
Use perPage variable instead of duplicated number.

ecgan committed:
Code refactor for Campaigns component, remove unneeded CampaignsCardHeader.

ecgan committed:
Add border and change padding for recommended channels in create campaign modal.

ecgan committed:
Merge branch 'feature/34903-multichannel-marketing-frontend/34905-campaigns-card' into feature/34903-multichannel-marketing-frontend/34909-create-campaign-modal

Conflicts:
	plugins/woocommerce-admin/client/marketing/hooks/index.ts
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.tsx
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Channels/Channels.tsx
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Channels/RecommendedChannels.tsx
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Channels/useChannels.ts

ecgan committed:
Merge branch 'feature/34905-marketing-campaigns-card' into feature/34909-marketing-create-campaign-modal

Conflicts:
	plugins/woocommerce-admin/client/marketing/components/RecommendedChannelsList.tsx
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.scss
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.tsx
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Channels/Channels.tsx
	plugins/woocommerce-admin/client/marketing/overview-multichannel/Channels/RecommendedChannels.tsx

ecgan committed:
Fix build error after merging.

ecgan committed:
wp.data for getting new campaign types.

ecgan committed:
Make UI work with wp.data new campaign types.

ecgan committed:
Merge branch 'feature/34905-marketing-campaigns-card' into feature/34909-marketing-create-campaign-modal

Conflicts:
	plugins/woocommerce-admin/client/marketing/data-multichannel/action-types.ts
	plugins/woocommerce-admin/client/marketing/data-multichannel/actions.ts
	plugins/woocommerce-admin/client/marketing/data-multichannel/resolvers.ts

ecgan committed:
Simplify wp.data for receiveCampaignTypes.

ecgan committed:
Simplify receiveCampaignTypes.

ecgan committed:
Move useNewCampaignTypes to shared hooks directory.

ecgan committed:
Load campaign types in Campaigns card.

ecgan committed:
Refetch registered channels after installing channel in Campaigns card.

ecgan commented:
@puntope , this PR is based on top of PR https://github.com/woocommerce/woocommerce/pull/36735, could you give this a review too please?

ecgan committed:
Add changelog.

ecgan committed:
Fix and add tests for Campaigns card.

ecgan committed:
Add tests for CreateNewCampaignModal.

ecgan committed:
Merge branch 'trunk' into feature/34909-marketing-create-campaign-modal

ecgan committed:
Fix array length rendering literal 0 in CreateNewCampaignModal.

eason9487 commented on plugins/woocommerce-admin/client/marketing/data-multichannel/resolvers.ts:
```diff
@@ -114,3 +116,19 @@ export function* getCampaigns( page: number, perPage: number ) {
+			yield receiveCampaignTypes( error );
+		}
+
+		throw error;
```
As I understood, errors thrown from resolvers won't be handled internally by wp-data and cannot be caught externally. Re-throwing will be an uncaught error and further cause the Campaigns card to keep getting stuck on the loading status.

![image](https://user-images.githubusercontent.com/17420811/223671376-602826f9-f9ab-41d4-96a0-4fe70baa96eb.png)

ecgan replied:
Hmm, that's weird, I tested it but the UI works as expected for me, it doesn't get stuck in loading status and I don't see uncaught error. 🤔 

![image](https://user-images.githubusercontent.com/417342/224034665-710592d2-3430-448d-9ccd-d7d534130e42.png)

Could it be due to the underlying WP environment? I'm using WP 6.1.1.

eason9487 replied:
> Could it be due to the underlying WP environment? I'm using WP 6.1.1.

I was using 5.9.5.

Just found the `@wordpress/data` resolvers started catching errors since [version 6.4.0](https://github.com/WordPress/gutenberg/blob/%40wordpress/data%406.4.0/packages/data/src/redux-store/index.js#L431-L452), and it's included in WP >= 6.0. See:
- https://github.com/WordPress/wordpress-develop/blob/5.9.5/package.json#L94
- https://github.com/WordPress/wordpress-develop/blob/6.0.0/package.json#L93

Considering [the latest WC still states it supports WP 5.9](https://github.com/woocommerce/woocommerce/blob/7.5.0-rc.1/plugins/woocommerce/woocommerce.php#L11) and this feature won't use `hasResolutionFailed` or `getResolutionError` to get the thrown errors, probably it would be better not to re-throw errors from the resolvers.

ecgan replied:
> Considering [the latest WC still states it supports WP 5.9](https://github.com/woocommerce/woocommerce/blob/7.5.0-rc.1/plugins/woocommerce/woocommerce.php?rgh-link-date=2023-03-10T07%3A59%3A56Z#L11) and this feature won't use `hasResolutionFailed` or `getResolutionError` to get the thrown errors, probably it would be better not to re-throw errors from the resolvers.

Yeah, that makes sense, to make it work with WP 5.9.

For reference, WP 5.9 uses `@wordpress/data@6.1.5`, and here's the corresponding code, there is no try/catch block: https://github.com/WordPress/gutenberg/blob/1081a28b2368fcebd9a14554fbd0e25eb33fbcb6/packages/data/src/redux-store/index.js#L414-L428

TIL about `hasResolutionFailed` or `getResolutionError`. Thanks for sharing that! Seems like it is introduced in PR https://github.com/WordPress/gutenberg/pull/38669. There should be more documentation around that. In the future we don't need to manually store the error on our own like what I did in this project. 😂 

I have removed the `throw error;` line for `getCampaignTypes` resolver in [`1fea018` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/1fea0185fe0002524bd7cf2a28dc915e8f9d0aa2). For other resolvers, I shall do it in separate PR.

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.tsx:
```diff
@@ -0,0 +1,164 @@
+					) }
+				</div>
+				<div>
+					{ newCampaignTypes?.map( ( el ) => {
```
It's minor but the visual result looks a bit weird when errors occurred in fetching campaign types.

![image](https://user-images.githubusercontent.com/17420811/223641754-c59d815d-e964-4fed-9422-545cbf0208be.png)

ecgan replied:
Good catch. Not just in error situation, it could also happen when a registered channel does not provide any campaign types (also related to https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1131035854).

I made a change in [`e4da8cb` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/e4da8cba0976ea83eafd86c7033b5bf2c2d406cd), to display a "No campaign types found" message when there are no campaign types.

![image](https://user-images.githubusercontent.com/417342/224118533-898e7e08-8dd2-4b10-92f2-6a79a6e38607.png)

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.tsx:
```diff
@@ -0,0 +1,164 @@
+									>
+										<Flex gap={ 1 }>
+											<FlexItem>
+												{ __(
+													'Create',
+													'woocommerce'
+												) }
```
I guess this is something that needs to be handled by the API side of Google Listings & Ads, but it's worth mentioning that this path is incorrect before the Google Ads setup has been completed in GLA.

https://user-images.githubusercontent.com/17420811/223647105-b82163b2-24a8-41bf-87e0-7cfcf5222623.mp4

If an extension is still listed in this place of the modal before its ad setup is completed, then I was thinking that maybe extensions like GLA could provide a consistent intermediary URL. The intermediary URL maintained by extensions decides which page to further navigate to depending on whether its ad setup is completed or not.

ecgan replied:
Good point!

> If **_an extension is still listed in this place of the modal_** before its ad setup is completed, then I was thinking that maybe extensions like GLA could provide a consistent intermediary URL. **_The intermediary URL_** maintained by extensions decides which page to further navigate to depending on whether its ad setup is completed or not.

An alternative that I can think of: if GLA ads setup is not completed yet, don't return the GLA ads campaign type in the get campaign types API, i.e. don't list it in this place of the modal. cc @nima-karimi , what do you think about this? The downside to this approach is that the users may not know about this until they go to GLA and finish the ads setup there.

I do agree that the intermediary URL is a better sensible approach though.

nima-karimi replied:
A dynamic intermediary URL set by each extension is the ideal solution. I will create an issue to implement this later.

nima-karimi replied:
Created an issue here: https://github.com/woocommerce/google-listings-and-ads/issues/1922

eason9487 commented on plugins/woocommerce-admin/client/marketing/hooks/useNewCampaignTypes.ts:
```diff
@@ -0,0 +1,46 @@
+		const campaignTypesState = getCampaignTypes< CampaignTypesState >();
+
+		return {
+			loading: ! hasFinishedResolution( 'getCampaignTypes' ),
```
Precisely, `! hasFinishedResolution` is not exactly equivalent to the loading status, although it works correctly here but is not very precise in the semantic meaning.

ecgan replied:
> is not very precise in the semantic meaning.

Do you have a suggestion? I'm guessing something like the following?

```suggestion
			hasFinishedResolution: hasFinishedResolution( 'getCampaignTypes' ),
```

And then replace the usage of `loading` with `! hasFinishedResolution`?

eason9487 replied:
Maybe `hasResolved`? There are some namings that are the same or start with it in the Gutenberg repo, and also a few in this repo. Another pattern like `fooResolved` might be okay as well.

ecgan replied:
Hmm, but would `hasResolved` be confused with `hasFinishedResolution` vs `hasStartedResolution`? 😂 

We can continue discussing about this naming. If we want to change it, I shall do it in a separate PR, to change all the `loading` to a name that we agree with. For now, I'll leave it as is.

eason9487 replied:
> but would `hasResolved` be confused with `hasFinishedResolution` vs `hasStartedResolution`?

`hasStartedResolution` is hardly used, and the semantics of *Resolved* is closer to *Finished* than *Started*, so it shouldn't be that easy to confuse. If this is still a concern, then there is a case for using `hasFinished` and `hasStarted` to distinguish between them.

[The exact same name `hasFinishedResolution`](https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1131043294) is not recommended, although it can be recognized by TS to be a boolean rather than a wp-data `hasFinishedResolution` selector, it is still relatively confusing when viewed on GH.

ecgan replied:
> `hasStartedResolution` is hardly used, and the semantics of _Resolved_ is closer to _Finished_ than _Started_, so it shouldn't be that easy to confuse. If this is still a concern, then there is a case for using `hasFinished` and `hasStarted` to distinguish between them.

Thanks for the ideas! I'll think about this a bit more and I can come up with a separate PR to change them later.

eason9487 commented on plugins/woocommerce-admin/client/marketing/hooks/useNewCampaignTypes.ts:
```diff
@@ -0,0 +1,46 @@
+			data: campaignTypesState.data?.map( convert ),
+			error: campaignTypesState.error,
+		};
+	} );
```
It would be a bit nicer to specify `deps` for `useSelect` in order not to return new references in every rerendering.

```suggestion
	}, [] );
```

ecgan replied:
Yeah, we should always specify `deps`, I missed that. I fixed it in [`3ba492f` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/3ba492f32b52e39650fc9a5d021279c2afc19689) (moving the `convert` function outside of the hook so it doesn't become a dependency) and in [`39ce470` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/39ce470fe8a3ed7370a46f0815307f7c2d6305da) (specifying empty dependency).

On a related note, ESLint should have caught these dependency issues, I don't know why it is not working for `useSelect`. It does work with normal React hooks like `useEffect`. Do you happen to know why?

eason9487 replied:
> Do you happen to know why?

Seems it couldn't distinguish whether a custom React hook is designed to use `deps` or not. Usually it needs to be specified separately. Ref: https://github.com/woocommerce/google-listings-and-ads/pull/780

ecgan replied:
Ahh got it! I think we should contribute it upstream to https://www.npmjs.com/package/@wordpress/eslint-plugin and https://www.npmjs.com/package/@woocommerce/eslint-plugin. I'll see if I can make it happen, maybe after this project. That should help fix or prevent a lot of these errors.

More info about the `"react-hooks/exhaustive-deps"` ESLint rule is in https://www.npmjs.com/package/eslint-plugin-react-hooks.

ecgan replied:
> I think we should contribute it upstream to [npmjs.com/package/@wordpress/eslint-plugin](https://www.npmjs.com/package/@wordpress/eslint-plugin)

I have created issue https://github.com/WordPress/gutenberg/issues/49897 and PR https://github.com/WordPress/gutenberg/pull/49900.

eason9487 commented on plugins/woocommerce-admin/client/marketing/hooks/useNewCampaignTypes.ts:
```diff
@@ -0,0 +1,46 @@
+	error?: ApiFetchError;
+};
+
+export const useNewCampaignTypes = (): UseNewCampaignTypes => {
```
I've read through a larger scope of code and design documents to try to understand more contexts but still unsure about what "New" means in the naming. Especially this hook doesn't seem to have any relevant things about "New", so I would suggest renaming it if not relevant.

ecgan replied:
The "New" is related to the "Create a new campaign" context, e.g. the UI modal. But I think I get your point though. I've renamed them in [`f19969d` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/f19969dacbe6c8511062101307a93f822958141c).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.test.tsx:
```diff
@@ -131,4 +155,30 @@ describe( 'Campaigns component', () => {
+
+		await userEvent.click(
+			screen.getByRole( 'button', { name: 'Create new campaign' } )
+		);
```
How about using `getByRole( 'button', ...` instead of `screen.getByText( 'Create new campaign' )` ?

```suggestion
		const button = screen.getByRole( 'button', {
			name: /create new campaign/i,
		} );

		expect( button ).toBeInTheDocument();

		await userEvent.click( button );
```

ecgan replied:
Ah yes, we should always use `getByRole` over `getByText` when possible, based on https://testing-library.com/docs/queries/about/#priority. I changed one line and missed the other line. 

Looking at it now, we actually don't need the `expect( screen.getByText( 'Create new campaign' ) ).toBeInTheDocument();` at all. When we do `getByRole`, if the element is not found, it will throw an error, so I have removed the line in [`551db21` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/551db21e9bfd1155afd8cbb64b70341ee3c6345f).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.test.tsx:
```diff
@@ -8,12 +8,21 @@ import userEvent from '@testing-library/user-event';
+} ) );
+
+jest.mock( './CreateNewCampaignModal', () => ( {
+	CreateNewCampaignModal: () => <div>Create a new campaign</div>,
```
Maybe use another text to let the querying target in the assertion below clearer as there is a button with the text "Create new campaign" in the document.

https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.test.tsx#L180-L182

ecgan replied:
Good point. I made it clearer in [`e831e89` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/e831e899979cdb674eda82db5f94bc7e03de7056).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.tsx:
```diff
@@ -50,14 +53,24 @@ const perPage = 5;
 export const Campaigns = () => {
 	const [ page, setPage ] = useState( 1 );
-	const { loading, data, meta } = useCampaigns( page, perPage );
+	const [ open, setOpen ] = useState( false );
```
Perhaps replacing `open` with a name that more clearly conveys the state.

ecgan replied:
Done in [`f5bb849` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/f5bb8490ea149c0123582e1fa38245676990b2ce).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.tsx:
```diff
@@ -0,0 +1,164 @@
+/**
+ * Props for CreateNewCampaignModal, which is based on Modal.Props.
+ *
+ * Modal's title and children props are omitted because they are specified within the component
+ * and not needed to be specified by the consumer.
+ */
+type CreateCampaignModalProps = Omit< Modal.Props, 'title' | 'children' >;
```
The `title` prop is still available because the order of the object spreading is the last one.

https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.tsx#L53-L54

ecgan replied:
When the component consumer try to specify the title for `CreateNewCampaignModal`, they should see compile error "Property 'title' does not exist", so this shouldn't be a major concern.

![image](https://user-images.githubusercontent.com/417342/224110228-cc182a82-e7f2-4ed4-b9c8-562a91dc1b79.png)

However, I made a change in [`c6dcee4` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/c6dcee426b447b08bef26807164d1477fd1059fd) anyway, to make it clearer that title and className is not meant to be overwritten.

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.tsx:
```diff
@@ -0,0 +1,164 @@
+											/>
+											{ idx !==
+												recommendedChannels.length -
+													1 && <CardDivider /> }
```
Wonder why it needs to render `CardDivider`s but not display them?

https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.scss#L31-L33

ecgan replied:
😂  That's because I was doing copy and paste from the Channels component. I think at one point it was a shared component (or maybe I wanted to make it a shared component), so there was a CardDivider and I hid it by using CSS. Thanks for catching this. I have removed them in [`e15305b` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/e15305b7c3db4e3c6b650bd71002faa82e18aec5).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.tsx:
```diff
@@ -0,0 +1,164 @@
+											<SmartPluginCardBody
+												plugin={ el }
+												onInstalledAndActivated={
+													refetch
```
A possible case could be I have fully installed Google Listings & Ads but deactivated it. After activating it via this modal, does this feature expect GLA to appear in the first list in this modal? Currently, it doesn't refetch the data of `getCampaignTypes`.

https://user-images.githubusercontent.com/17420811/223702481-5aeb6ec1-44f5-4a2c-ba26-8c938c123b5e.mp4

ecgan replied:
Good catch! 

I made the fix in [`8b6aad4` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/8b6aad4138661e5a98386a67514adbba8e2bbeb7) for this CreateNewCampaignModal.

I also made the fix in [`6446053` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/644605378493bd649de8292292d9504ca9b6032a). When users activate a channel in the Channels card, we should also refetch the campaign types.

---

Food for thought / for future improvement: right now we are calling `refetchCampaignTypes` and `refetchRegisteredChannels` in an event handler in the above two places. I think it might be better if we come up with a wp.data action that contains these, e.g. `install channel -> activate channel -> invalidate resolution for campaign types and registered channels`.

eason9487 replied:
> I made the fix in [`8b6aad4` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/8b6aad4138661e5a98386a67514adbba8e2bbeb7) for this CreateNewCampaignModal.

After the 8b6aad4, it still won't refetch the data of `getCampaignTypes`.

The call to `invalidateResolution` in `useCampaignTypes` requires the second parameter:

```js
invalidateResolution( 'getCampaignTypes', [] );
```

eason9487 replied:
The function signature of `invalidateResolution` has been changed in WP version >= 6.0
- In WP version < 6.0, the second parameter has no default value
- In WP version >= 6.0, the second parameter defaults by `[]`

Therefore, with a WP version < 6.0, it cannot trigger state invalidation because of the mismatched `args` of a wp-data selector.

Ref:
- WP version < 6.0
   - https://github.com/WordPress/gutenberg/blob/%40wordpress/data%406.2.1/packages/data/src/redux-store/metadata/reducer.ts#L53-L57
- WP version >= 6.0
   - https://github.com/WordPress/gutenberg/blob/%40wordpress/data%406.3.0/packages/data/src/redux-store/metadata/reducer.ts#L56-L60
   - https://github.com/WordPress/gutenberg/blob/%40wordpress/data%406.3.0/packages/data/src/redux-store/metadata/utils.ts#L48-L51

ecgan replied:
Got it. Similar to https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1132055563, we need to make it work for WP 5.9, so I have made the fix in [`c6e6f87` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/c6e6f8774d6ec36f0d7ebe2c3e17e50d5c847e7c). I specified empty array dependency `[]` not just for the `invalidateResolution` but also for `hasFinishedResolution`, to make both of them consistent.

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/CreateNewCampaignModal/CreateNewCampaignModal.test.tsx:
```diff
@@ -0,0 +1,151 @@
+					channelName: 'Google Listings and Ads',
+					channelSlug: 'google-listings-and-ads',
+				},
+			],
```
Looks like this data is not relevant to this test case. Perhaps setting it to an empty array should be fine.

ecgan replied:
I thought about that. If I made it into an empty array, it might be a bit confusing because the line after it also uses empty array for `useRecommendedChannels`, so there would be two empty arrays for the test. I thought it would be clearer if one array is empty and one is not.

I guess your concern is about the big bunch of text hurting readability a bit. I made a refactor in [`29d67a3` (#37044)](https://github.com/woocommerce/woocommerce/pull/37044/commits/29d67a3a21d6ee5441218c0f611aef092380abab), I think it looks much better now.

eason9487 replied:
Mainly because that data is not relevant to this test objective.

> I thought about that. If I made it into an empty array, it might be a bit confusing because the line after it also uses empty array for `useRecommendedChannels`, so there would be two empty arrays for the test. I thought it would be clearer if one array is empty and one is not.

Got it. Making the dummy data share is also good. Thanks!

eason9487 reviewed (commented):
Thanks for the work.

There are some issues related to the management of data and state across React hooks and wp-data store. The rest are minor things.

---

In order to understand the context of this PR, during the code review I looked at a wider range of code and found some additional issues. They were not developed in this PR but I thought it's worth bringing them up:

1. It will show a "0" when no campaigns.
   - https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/Campaigns.tsx#L185
   ![image](https://user-images.githubusercontent.com/17420811/223672358-2695e6e7-d41a-44c9-9682-21dd076761ef.png)
2. Same as the `getCampaignTypes` reducer, the other three reducers also have the same concern of re-throwing errors.
3. Although not causing actual problems at this moment, there is a concern about incorrect state caching within `useCampaigns`, which could lead to incorrect results in the future for standalone components like `Campaigns`.
   - `channels` was missed  in the [deps of `useSelect`](https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/Campaigns/useCampaigns.ts#L83).
   - The potential problems could be simulated by returning the following code in the `MarketingOverviewMultichannel` component:
      ```js
      return (
      	<div className="woocommerce-marketing-overview-multichannel">
      		<Campaigns />
      	</div>
      );
      ```
   - The leading icons didn't appear:
      ![image](https://user-images.githubusercontent.com/17420811/223717960-ee0ab991-364d-4d19-9482-ae4ba947b718.png)
4. After installing or activating an extension in `MarketingOverviewMultichannel`, the [`refetch`](https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/MarketingOverviewMultichannel.tsx#L54) doesn't trigger the data refetching for `getRegisteredChannels`.
   - The current result:

      https://user-images.githubusercontent.com/17420811/223719204-b6fdda37-8290-488b-b00d-a68992b3d342.mp4
   - I guess the expected result is like this:

      https://user-images.githubusercontent.com/17420811/223719243-3227ba6b-89d4-48fc-93c9-d3b45a626fb2.mp4
5. In the relevant processing of `getCampaigns` in wp-data store and `useCampaigns` hook, they don't include `perPage` as a part of the data key. It will lead to incorrect data and states if other implementations use a different `perPage` to request data someday.
   - For example, when having more than 5 GLA campaigns, adding a line of code `useCampaigns( page, 10 );` to the `Campaigns` component could result in incorrect pagination.

      https://user-images.githubusercontent.com/17420811/223723790-bd981c12-431f-4816-868b-8afc49f393fd.mp4
6. After activating GLA via the Channels card, the GLA item in the Installed extensions card is still remaining unactivated. 

   https://user-images.githubusercontent.com/17420811/223721770-cf752b50-1a49-44af-bd00-49a7cf9d4d8c.mp4

ecgan committed:
Move convert function out of useSelect in useNewCampaignTypes.

This is so that it doesn't become dependency in useSelect.

ecgan committed:
Specify dependency for useSelect in useNewCampaignTypes.

ecgan committed:
Change NewCampaignTypes to CampaignTypes.

ecgan committed:
Remove unneeded getByText in Campaigns test.

ecgan committed:
Make the mocked modal clearer.

The previous text "Create a new campaign" can be a bit confusing with the other text "Create new campaign" in the test.

ecgan committed:
Rename open and setOpen to isModalOpen and setModalOpen in Campaigns.

ecgan committed:
Make it clear that title and classname cannot be overwritten by props.

ecgan committed:
Remove unneeded CardDivider in CreateNewCampaignModal.

The code is copied from the Channels component, but in CreateNewCampaignModal, we don't need to display the divider, so we can just remove the divider code.

ecgan committed:
Remove unneeded return in CreateNewCampaignModal.

This helps us save indentation and make the code few lines shorter.

ecgan committed:
Refactor test data in CreateNewCampaignModal tests.

ecgan committed:
Refetch campaign types after installing a channel in CreateNewCampaignModal.

ecgan committed:
Load campaign types in overview and refetch after installeing a channel.

ecgan committed:
Display info message when there are no campaign types.

ecgan committed:
Remove unneeded div in CreateNewCampaignModal.

ecgan commented:
@eason9487 , thanks for the thorough review, great eyes as always! 😀 👍 I have addressed your code review comments in this PR, this is ready for another round of review. 🙏 

----

The following is about the 6 points mentioned in your code review summary in https://github.com/woocommerce/woocommerce/pull/37044#pullrequestreview-1330049691. Thanks so much for bringing them up!

> 1. It will show a "0" when no campaigns.

I'll fix it in a separate PR. I'm thinking to check through the marketing code base to make sure this is handled correctly. We recently just fixed a similar issue https://github.com/woocommerce/woocommerce/issues/37083.

> 2. Same as the `getCampaignTypes` reducer, the other three reducers also have the same concern of re-throwing errors.

This is related to https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1129033825. The problem is I don't see the issue that you are seeing. We can talk about it in that comment thread.

> 3. Although not causing actual problems at this moment, there is a concern about incorrect state caching within `useCampaigns`, which could lead to incorrect results in the future for standalone components like `Campaigns`.

Agree. I actually thought of doing some changes with `useCampaigns`. I feel like `useCampaigns` is doing too much with `useRegisteredChannels` in it, I want to make it leaner and remove "channel" things. I'm thinking to work with what we have now and change it later.

> 4. After installing or activating an extension in `MarketingOverviewMultichannel`, the [`refetch`](https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/MarketingOverviewMultichannel.tsx#L54) doesn't trigger the data refetching for `getRegisteredChannels`.

I couldn't reproduce this, it works as expected for me. 🤔 See my demo video below.

https://user-images.githubusercontent.com/417342/224124936-82074979-96b4-4289-8dfd-20b286cce1a9.mp4

> 5. In the relevant processing of `getCampaigns` in wp-data store and `useCampaigns` hook, they don't include `perPage` as a part of the data key. It will lead to incorrect data and states if other implementations use a different `perPage` to request data someday.

Good point! I guess I can make the `perPage` as a part of the key, or don't expose `perPage` as a parameter 😂 . I'll create a separate issue and label it as low priority since this isn't a major issue now.

> 6. After activating GLA via the Channels card, the GLA item in the Installed extensions card is still remaining unactivated.

This should not be a problem with issue https://github.com/woocommerce/woocommerce/issues/34907 and its PR https://github.com/woocommerce/woocommerce/pull/37126, because the extension should only appear in one card, it shouldn't appear in both Channels card and Installed extensions card.

eason9487 commented:
> > 4\. After installing or activating an extension in `MarketingOverviewMultichannel`, the [`refetch`](https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/MarketingOverviewMultichannel.tsx#L54) doesn't trigger the data refetching for `getRegisteredChannels`.
> 
> I couldn't reproduce this, it works as expected for me.

Just realized it's because of different WP versions. See https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1132062493 and https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1132097171.

Considering [the latest WC still states it supports WP 5.9](https://github.com/woocommerce/woocommerce/blob/7.5.0-rc.1/plugins/woocommerce/woocommerce.php#L11), I guess it would still need to add the second parameter to the `invalidateResolution` call in `useRegisteredChannels` at this moment.

eason9487 reviewed (commented):
Thanks for the updates.

The two remaining concerns are about the compatibility with WP 5.9:
- https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1129033825
- https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1129226234

About the remaining issues mentioned in the main content of this https://github.com/woocommerce/woocommerce/pull/37044#pullrequestreview-1330049691, I also think they could be addressed in separate PRs. 👍

ecgan committed:
Don't throw error in getCampaignTypes.

In WP 5.9, if we throw error, there would be an uncaught promise, and it causes UI to break.

ecgan committed:
Specify empty array dependency in useCampaignTypes.

This is to make it work with WP 5.9. See https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1132062493.

ecgan commented:
> > > 4. After installing or activating an extension in `MarketingOverviewMultichannel`, the [`refetch`](https://github.com/woocommerce/woocommerce/blob/97ed1d2a7d25a6511535e60edfb8edc26b066cc2/plugins/woocommerce-admin/client/marketing/overview-multichannel/MarketingOverviewMultichannel.tsx#L54) doesn't trigger the data refetching for `getRegisteredChannels`.
> > 
> > 
> > I couldn't reproduce this, it works as expected for me.
> 
> Just realized it's because of different WP versions. See [#37044 (comment)](https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1132062493) and [#37044 (comment)](https://github.com/woocommerce/woocommerce/pull/37044#discussion_r1132097171).
> 
> Considering [the latest WC still states it supports WP 5.9](https://github.com/woocommerce/woocommerce/blob/7.5.0-rc.1/plugins/woocommerce/woocommerce.php?rgh-link-date=2023-03-10T08%3A53%3A28Z#L11), I guess it would still need to add the second parameter to the `invalidateResolution` call in `useRegisteredChannels` at this moment.

Got it. I have created a separate issue https://github.com/woocommerce/woocommerce/issues/37176 for this.

> > 1. It will show a "0" when no campaigns.
> 
> I'll fix it in a separate PR. I'm thinking to check through the marketing code base to make sure this is handled correctly. We recently just fixed a similar issue #37083.

Created issue https://github.com/woocommerce/woocommerce/issues/37175. 

> > 9\. In the relevant processing of `getCampaigns` in wp-data store and `useCampaigns` hook, they don't include `perPage` as a part of the data key. It will lead to incorrect data and states if other implementations use a different `perPage` to request data someday.
> 
> Good point! I guess I can make the `perPage` as a part of the key, or don't expose `perPage` as a parameter 😂 . I'll create a separate issue and label it as low priority since this isn't a major issue now.

Created issue https://github.com/woocommerce/woocommerce/issues/37177.

----

@eason9487 , all the code review comments are addressed. This is ready for another round of review. 🙂 🙏

eason9487 reviewed (approved):
Thanks for the further adjustments! LGTM.

ecgan committed:
Merge branch 'trunk' into feature/34909-marketing-create-campaign-modal
